### ISSUE TITLE
How do we handle compiler warnings with gcc9 for the 9.1 release?

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#8134]
As indicated in #8015 we did have a series of warnings with gcc9. After the fix #8034 I started to look into the issues reported via bundled software. Most of the warnings are of this kind (see also [1] for 300 of those warnings), now the one I found in our code:

```
/.../dealii_git/source/lac/petsc_parallel_sparse_matrix.cc: In member function ‘PetscScalar dealii::PETScWrappers::MPI::SparseMatrix::matrix_norm_square(const dealii::PETScWrappers::MPI::Vector&) const’:
/.../dealii_git/source/lac/petsc_parallel_sparse_matrix.cc:679:19: warning: implicitly-declared ‘dealii::PETScWrappers::MPI::Vector::Vector(const dealii::PETScWrappers::MPI::Vector&)’ is deprecated [-Wdeprecated-copy]
  679 |       Vector tmp(v);

/.../dealii_git/include/deal.II/lac/petsc_vector.h:277:7: note: because ‘dealii::PETScWrappers::MPI::Vector’ has user-provided ‘dealii::PETScWrappers::MPI::Vector& dealii::PETScWrappers::MPI::Vector::operator=(const dealii::PETScWrappers::MPI::Vector&)’
  277 |       operator=(const Vector &v);
      |       ^~~~~~~~
```

The warnings go both ways, namely complaints about classes that have a copy constructor but no copy assignment, and classes that have a copy assignment but not copy constructor. Many of them are located in the bundled boost-1.62.0. There is one in muparser and some in TBB (`enumerable_thread_specific`).

The question is how we handle this kind of warning for the release. My suggestion would be to disable the respective gcc warning `-Wno-deprecated-copy`. We could potentially fix them (I have boost and muparser things addressed locally, but not yet TBB).

The second kind of warning, which I have not yet addressed, is 
```
/home/kronbichler/sw/dealii_git/bundled/boost-1.62.0/include/boost/geometry/index/detail/varray_detail.hpp:544:13: warning: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type ‘struct std::pair<dealii::Point<3, double>, unsigned int>’ with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
  544 |     ::memcpy(boost::addressof(*pos), boost::addressof(v), sizeof(V));
```
This is because we end up in a path for `memcpy` when we have a `std::pair` of `dealii::Point,uint` that is not trivially copyable. Again, we could probably address this by modifications in boost, but it is not trivial and I have not yet found where boost decides that the type in question would be trivially copyable.

As we have discussed during the last release in #6618, we might need to live with some warnings for new compilers. Yet, the situation is a bit frustrating due to the huge numbers of warnings that hides any useful error message, so I guess some action would help.

[1] https://cdash.kyomu.43-1.org/viewBuildError.php?type=1&buildid=9756

**Comment by drwells** [comment]
If we can resolve these by writing, e.g., `Vector(const Vector &) = default` then I think we should do that. If its anything more complex then we should not. I would not worry about the bundled libraries: it might just be the way I operate, but anyone who can install the latest compiler can usually also install the latest libraries so that case is less significant.

**Comment by drwells** [comment]
I will take a quick look at this in an hour or so.

**Comment by kronbichler** [comment]
@drwells I have fixes for PETSc and boost already.

**Comment by kronbichler** [comment]
I can post them (as separate patches), for whatever we decide.

**Comment by drwells** [comment]
that was quick :)

**Comment by tamiko** [comment]
In addition / Alternatively, I suggest we add the corresponding `-Wno-*` flag to `DEAL_II_DISABLE_EXTRA_DIAGNOSTICS`.

**Comment by kronbichler** [comment]
Yes, the solution #8139 is what I would prefer as well (I just already had the changes halfway done, so I posted them in #8136, for the better or worse).

**Comment by kronbichler** [comment]
Closed by #8139.
